lib/breakers/service.rb

```diff
@@ -0,0 +1,110 @@
+require 'uri'
+
+module Breakers
+  class Service
+    attr_reader :name
+    attr_reader :evaluator
+
+    ONE_MONTH = 60 * 60 * 24 * 30
+
+    def initialize(name:, &block)
+      @name = name
+      @evaluator = block
+    end
+
+    def add_error
+      increment_key(key: errors_key)
+      maybe_create_outage
+    end
+
+    def add_success
+      increment_key(key: successes_key)
+    end
+
+    def last_outage
+      Outage.find_last(service: self)
+    end
+
+    def outages_in_range(start_time:, end_time:)
+      Outage.in_range(
+        service: self,
+        start_time: start_time,
+        end_time: end_time
+      )
+    end
+
+    def successes_in_range(start_time:, end_time:)
+      values_in_range(start_time: start_time, end_time: end_time, type: :successes)
+    end
+
+    def errors_in_range(start_time:, end_time:)
+      values_in_range(start_time: start_time, end_time: end_time, type: :errors)
+    end
+
+    def uri_name
+      URI.escape(name)
+    end
+
+    protected
+
+    def errors_key(time: nil)
+      "cb-#{name}-errors-#{align_time_on_minute(time: time).to_i}"
+    end
+
+    def successes_key(time: nil)
+      "cb-#{name}-successes-#{align_time_on_minute(time: time).to_i}"
+    end
+
+    def values_in_range(start_time:, end_time:, type:, sample_seconds: 3600)
+      start_time = align_time_on_minute(time: start_time)
+      end_time = align_time_on_minute(time: end_time)
+      keys = []
+      times = []
+      while start_time < end_time
+        times << start_time
+        if type == :errors
+          keys << errors_key(time: start_time)
+        elsif type == :successes
+          keys << successes_key(time: start_time)
+        end
+        start_time += sample_seconds
+      end
+      Breakers.client.redis_connection.mget(keys).each_with_index.map do |value, idx|
+        { count: value.to_i, time: times[idx] }
+      end
+    end
+
+    def increment_key(key:)
+      Breakers.client.redis_connection.multi do
+        Breakers.client.redis_connection.incr(key)
+        Breakers.client.redis_connection.expire(key, ONE_MONTH)
+      end
+    end
+
+    # Take the current or given time and round it down to the nearest minute
+    def align_time_on_minute(time: nil)
+      time = (time || Time.now).to_i
+      time - (time % 60)
+    end
+
+    def maybe_create_outage
+      data = Breakers.client.redis_connection.multi do
+        Breakers.client.redis_connection.get(errors_key(time: Time.now))
+        Breakers.client.redis_connection.get(errors_key(time: Time.now - 60))
+        Breakers.client.redis_connection.get(successes_key(time: Time.now))
+        Breakers.client.redis_connection.get(successes_key(time: Time.now - 60))
```

Comment:
With building this as a gem, using Time.now vc for example Time.current (ActiveSupport) wouldn't this maybe cause issues? Should it at least maybe always be .utc ?